source/common/router/router.cc

```diff
@@ -914,6 +914,34 @@ void Filter::onPerTryTimeout(UpstreamRequest& upstream_request) {
                          StreamInfo::ResponseCodeDetails::get().UpstreamPerTryTimeout);
 }
 
+void Filter::onStreamMaxDurationReached(UpstreamRequest& upstream_request) {
+  upstream_request.resetStream();
+
+  // In general, we don't have to retry when downstream didn't end. But when stream max duration
+  // timer was invoked, We must retry with keeping downstream connection so that we must allow to
+  // retry by switching full streaming retry flag.
+  allow_full_streaming_retry_ = true;
```

Comment:
I don't want to change this behavior in this PR as it is very complicated and @ggreenway is working on this in https://github.com/envoyproxy/envoy/pull/10725. So I would either revert this (I think it's fine to just handle the basic case where downstream did end but duration still fires) or let's wait for that PR to land.

Reply:
I see. ok. For now, I stop to work this and wait to resolve #10725. Can you add nostale flag to this PR?